# 1. 31

## 1.1. 问题

### 1.1.1. 下面列举的是 recover() 的几种调用方式，哪些是正确的？

A.

```go
func main() {
  recover()
  panic(1)
}
```

B.

```go
func main() {
  defer recover()
  panic(1)
}
```

C.

```go
func main() {
  defer func() {
      recover()
  }()
  panic(1)
}
```

D.

```go
func main() {
  defer func() {
      defer func() {
         recover()
      }()
  }()
  panic(1)
}
```

### 1.1.2. 下面代码输出什么，请说明？

```go
func main() {
    defer func() {
        fmt.Print(recover())
    }()
    defer func() {
        defer fmt.Print(recover())
        panic(1)
    }()
    defer recover() 
    panic(2)
}
```


### 1.1.3. 下面的代码输出什么，请说明？

```go
func main() {
    defer func() {
        fmt.Print(recover())
    }()
    defer func() {
        defer func() {
            fmt.Print(recover())
        }()
        panic(1)
    }()
    defer recover()
    panic(2)
}
```


## 1.2. 答案

### 1.2.1. 答案1

参考答案及解析：C。

**`recover() 必须在 defer() 函数中直接调用才有效`**。

**直接调用 recover()、在 defer() 中直接调用 recover() 和 defer() 调用时多层嵌套，这几种情况都是无效的。**



### 1.2.2. 答案2

参考答案及解析：21。

**recover() 必须在 defer() 函数中调用才有效，所以第 9 行代码捕获是无效的。**

**在调用 defer() 时，便会计算函数的参数并压入栈中，所以执行第 6 行代码时，此时便会捕获 panic(2)；**

**此后的 panic(1)，会被上一层的 recover() 捕获。**

所以输出 21。

### 1.2.3. 答案3

参考答案及解析：12。

注意：`defer recover()` 这样是不规范的，GoLand 直接提示 `defer 不应直接调用 recover()`
